File: system-test/typescript/src/testkit/dsl/port/myShop/given/steps/given-clock.ts
import type { GivenStage } from '../given-stage.js';
import type { WhenStage } from '../../when/when-stage.js';
import type { ThenStage } from '../../then/then-stage.js';

export interface GivenClock {
  withTime(time?: string): GivenClock;
  withWeekday(): GivenClock;
  withWeekend(): GivenClock;
  and(): GivenStage;
  when(): WhenStage;
  then(): ThenStage;
}

File: system-test/dotnet/Dsl.Core/Scenario/Given/Steps/GivenClock.cs
using Dsl.Port.Given.Steps;
using Driver.Adapter;
using static Dsl.Core.Scenario.ScenarioDefaults;

namespace Dsl.Core.Scenario.Given;

public class GivenClock : BaseGiven, IGivenClock
{
    private string? _time;

    public GivenClock(GivenStage givenClause) : base(givenClause)
    {
        WithTime(DefaultTime);
    }

    public GivenClock WithTime()
    {
        return WithTime(DefaultTime);
    }

    public GivenClock WithTime(string? time)
    {
        _time = time;
        return this;
    }

    IGivenClock IGivenClock.WithTime() => WithTime();
    IGivenClock IGivenClock.WithTime(string? time) => WithTime(time);

    public GivenClock WithWeekday()
    {
        return WithTime(WeekdayTime);
    }

    IGivenClock IGivenClock.WithWeekday() => WithWeekday();

    public GivenClock WithWeekend()
    {
        return WithTime(WeekendTime);
    }

    IGivenClock IGivenClock.WithWeekend() => WithWeekend();

    internal override async Task Execute(UseCaseDsl app)
    {
        (await app.Clock().ReturnsTime()
            .Time(_time)
            .Execute())
            .ShouldSucceed();
    }
}


